Source/Core/InputCommon/ControllerInterface/CemuHookUDPServer/CemuHookUDPServerProto.h

```diff
@@ -0,0 +1,262 @@
+  u32 pad_request_count;
+  u8 pad_id[4];
+};
+
+struct PortInfo
+{
+  static constexpr auto FROM = SERVER;
+  static constexpr auto TYPE = 0x100001U;
+  MessageHeader header;
+  u32 message_type;
+  u8 pad_id;
+  DsState pad_state;
+  DsModel model;
+  DsConnection connection_type;
+  u8 pad_mac_address[6];
+  DsBattery battery_status;
+  u8 padding;
+};
+
+struct PadDataRequest
+{
+  static constexpr auto FROM = CLIENT;
+  static constexpr auto TYPE = 0x100002U;
+  MessageHeader header;
+  u32 message_type;
+  RegisterFlags register_flags;
+  u8 pad_id_to_register;
+  u8 mac_address_to_register[6];
+};
+
+struct PadDataResponse
+{
+  static constexpr auto FROM = SERVER;
+  static constexpr auto TYPE = 0x100002U;
+  MessageHeader header;
+  u32 message_type;
+  u8 pad_id;
+  DsState pad_state;
+  DsModel model;
+  DsConnection connection_type;
+  u8 pad_mac_address[6];
+  DsBattery battery_status;
+  u8 active;
+  u32 hid_packet_counter;
+  u8 button_states1;
+  u8 button_states2;
+  u8 button_ps;
+  u8 button_touch;
+  u8 left_stick_x;
+  u8 left_stick_y_inverted;
+  u8 right_stick_x;
+  u8 right_stick_y_inverted;
+  u8 button_dpad_left_analog;
+  u8 button_dpad_down_analog;
+  u8 button_dpad_right_analog;
+  u8 button_dpad_up_analog;
+  u8 button_square_analog;
+  u8 button_cross_analog;
+  u8 button_circle_analog;
+  u8 button_triangle_analog;
+  u8 button_r1_analog;
+  u8 button_l1_analog;
+  u8 trigger_r2;
+  u8 trigger_l2;
+  Touch touch1;
+  Touch touch2;
+  u64 timestamp_us;
+  float accelerometer_x_g;
+  float accelerometer_y_g;
+  float accelerometer_z_inverted_g;
+  float gyro_pitch_deg_s;
+  float gyro_yaw_deg_s;
+  float gyro_roll_deg_s;
+};
+
+struct FromServer
+{
+  union
+  {
+    struct
+    {
+      MessageHeader header;
+      u32 message_type;
+    };
+    MessageType::VersionResponse version_response;
+    MessageType::PortInfo port_info;
+    MessageType::PadDataResponse pad_data_response;
+  };
+};
+
+struct FromClient
+{
+  union
+  {
+    struct
+    {
+      MessageHeader header;
+      u32 message_type;
+    };
+    MessageType::VersionRequest version_request;
+    MessageType::ListPorts list_ports;
+    MessageType::PadDataRequest pad_data_request;
+  };
+};
+}  // namespace MessageType
+
+static inline u32 CRC32(const void* buffer, unsigned length)
+{
+  return crc32(crc32(0L, Z_NULL, 0), (const Bytef*)buffer, length);
+}
+
+template <typename MsgType>
+struct Message : public MsgType
+{
+  Message() {}
+
+  Message(u32 source_uid)
+  {
+    memset(this, 0, sizeof *this);
+    static_assert(sizeof(char) == sizeof(u8));
+    strncpy((char*)MsgType::header.source, MsgType::FROM, sizeof MsgType::header.source);
+    MsgType::header.protocol_version = CEMUHOOK_PROTOCOL_VERSION;
+    MsgType::header.message_length = sizeof *this - sizeof MsgType::header;
+    MsgType::header.source_uid = source_uid;
+    MsgType::message_type = MsgType::TYPE;
+  }
+
+  void Finish() { MsgType::header.crc32 = CRC32(this, sizeof *this); }
```

Comment:
Ditto for the same reason. You'd be explicitly passing the data to calculate the CRC of, rather than the class itself.

Reply:
I decided to stick with inheritance. See other conversation about `memset(this, 0, sizeof *this);` for my reasons.

Reply:
The issue I see with using `this` like in the `CRC32` call is that it requires there to be only one struct/class in the inheritance chain with non static data members. Documenting this in a comment for potential future developers should solve the issue, shouldn't it?